image_view/src/nodes/image_saver.cpp

```diff
@@ -38,35 +38,127 @@
 #include <image_transport/image_transport.h>
 #include <camera_calibration_parsers/parse.h>
 #include <boost/format.hpp>
+#include <boost/thread.hpp>
 
 #include <std_srvs/Empty.h>
 #include <std_srvs/Trigger.h>
 
-boost::format g_format;
-bool save_all_image, save_image_service;
-std::string encoding;
-bool request_start_end;
-
-
-bool service(std_srvs::Empty::Request &req, std_srvs::Empty::Response &res) {
-  save_image_service = true;
-  return true;
-}
 
 /** Class to deal with which callback to call whether we have CameraInfo or not
  */
 class Callbacks {
 public:
-  Callbacks() : is_first_image_(true), has_camera_info_(false), count_(0) {
+  Callbacks(ros::NodeHandle& nh,
+            const std::string& topic,
+            const std::string& filename_format,
+            const std::string& encoding,
+            const bool save_all_image,
+            const bool wait_for_save,
+            const double save_timeout)
+      : nh_(nh), it_(nh_), topic_(topic), encoding_(encoding),
+        count_(0), start_time_(0),
+        subscribing_(false), should_unsubscribe_(false),
+        wait_for_save_(wait_for_save), save_timeout_(save_timeout)
+  {
+    format_.parse(filename_format);
+    if (save_all_image)
+    {
+      subscribe();
+    } else {
+      srv_save_ = nh_.advertiseService("save", &Callbacks::callbackSave, this);
+      srv_start_ = nh_.advertiseService("start", &Callbacks::callbackStartSave, this);
+      srv_end_ = nh_.advertiseService("end", &Callbacks::callbackEndSave, this);
+    }
   }
 
-  bool callbackStartSave(std_srvs::Trigger::Request &req,
-                         std_srvs::Trigger::Response &res)
+  void subscribe()
   {
-    ROS_INFO("Received start saving request");
+    boost::mutex::scoped_lock lock(mutex_);
```

Comment:
Please use `std::lock_guard` with a `std::mutex`.

Reply:
In order to build C++11, we need to set C++11 flag in CMakelists.txt.
I update this PR to add this change https://github.com/ros-perception/image_pipeline/pull/412/commits/cf0576a8436f2cfcd5c4f1343fe67dac7fac2d1a